cpp/src/arrow/util/thread_pool_test.cc

```diff
@@ -452,6 +454,42 @@ TEST_F(TestThreadPool, QuickShutdown) {
   add_tester.CheckNotAllComputed();
 }
 
+TEST_F(TestThreadPool, DestroyWithoutShutdown) {
+  std::mutex mx;
+  std::condition_variable cv;
+  std::weak_ptr<ThreadPool> weak_pool;
+  bool ready = false;
+  bool completed = false;
+  {
+    auto pool = this->MakeThreadPool(1);
+    weak_pool = pool;
+    // Simulate Windows
+    pool->shutdown_on_destroy_ = false;
+
+    ASSERT_OK(pool->Spawn([&mx, &cv, &completed, &ready] {
+      std::unique_lock<std::mutex> lock(mx);
+      ready = true;
+      cv.notify_one();
```

Comment:
Hmm... how do you know the notified one won't be the `wait` call just below?

Also, I'm not sure that `notify_one` *guarantees* that only one thread will be notified. For example, the [POSIX equivalent](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cond_signal.html) says:
> The `pthread_cond_signal()` function shall unblock *at least one* of the threads that are blocked on the specified condition variable

(emphasis mine)

Reply:
I'm pretty sure you're right, notify may wake up more than one thread.  However, I don't think it can wake up a thread before it starts waiting.  In other words, a thread can never notify itself.

I can probably rewrite this with two conditions though to make it more legible and guard against spurious wakeups.  My goal here is to ensure that the worker thread is spawned before I lose the reference to the thread pool.  Although, thinking on this again, the synchronization may not be needed at all since the lambda that creates a copy of the thread pool's shared_ptr is guaranteed to have been created before `Spawn` returns.

Reply:
Yes, you should probably use two different predicates and booleans.

Reply:
I've cleaned this test up to use two predicates & booleans.